src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,264 @@
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class Virtualize<TItem> : ComponentBase, IVirtualizeJsCallbacks, IAsyncDisposable
+    {
+        private readonly List<TItem> _loadedItems = new List<TItem>();
+
+        private VirtualizeJsInterop? _jsInterop;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        private int _itemsBefore;
+
+        private int _visibleItemCapacity;
+
+        private int _itemsVisible;
+
+        private int _itemCount;
+
+        private Task _fetchTask = Task.CompletedTask;
+
+        private CancellationTokenSource? _fetchCts;
+
+        private ItemsProviderDelegate<TItem> _itemsProvider = default!;
+
+        private RenderFragment<TItem>? _itemTemplate;
+
+        private RenderFragment<int>? _placeholder;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? ChildContent { get; set; }
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem>? Item { get; set; }
+
+        /// <summary>
+        /// Gets or sets the template for items that have not yet been loaded in memory.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<int>? Placeholder { get; set; }
+
+        /// <summary>
+        /// Gets the size of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the fixed item source.
+        /// </summary>
+        [Parameter]
+        public ICollection<TItem> Items { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (ItemSize <= 0)
+            {
+                throw new InvalidOperationException(
+                    $"{GetType()} requires a positive value for parameter '{nameof(ItemSize)}' to perform virtualization.");
+            }
+
+            if (ItemsProvider != null)
+            {
+                if (Items != null)
+                {
+                    throw new InvalidOperationException($"{GetType()} can only accept one item source from its parameters.");
```

Comment:
Just to reduce any possible confusion:

```suggestion
                    throw new InvalidOperationException($"{GetType()} can only accept one item source from its parameters. Do not supply both {nameof(Items)} and {nameof(ItemsProvider)}.");
```